custom/icds/location_reassignment/download.py

```diff
@@ -1,10 +1,129 @@
+        """
+        Generates an Excel file stream
+        With details of all locations related to a location
+        and users assigned to these locations.
+        Each sheet corresponds to a location type.
+        """
+        self.location = SQLLocation.active_objects.get(location_id=location_id)
 
     def dump(self):
+        self._init_location_details()
+        self._populate_assigned_users()
+        wb = self._create_workbook()
         stream = io.BytesIO()
+        wb.save(stream)
         return stream
+
+    def _init_location_details(self):
+        # setup details for the main location and all related locations
+        self._location_details_by_location_id = {}
+        for location in self._locations():
+            self._location_details_by_location_id[location.location_id] = {
+                'name': location.name,
+                'site_code': location.site_code,
+                'lgd_code': location.metadata.get('lgd_code', ''),
+                'parent_location_id': location.parent_location_id,
+                'location_type': location.location_type.code,
+                'assigned_users': [],
+            }
+        return self._location_details_by_location_id
+
+    def _locations(self):
+        # fetch all locations necessary for this download request
+        ancestors = list(self.location.get_ancestors().select_related('location_type'))
+        self_and_descendants = list(self.location.get_descendants(include_self=True)
+                                    .select_related('location_type'))
+        return ancestors + self_and_descendants
+
+    def _populate_assigned_users(self):
+        # allot assign user details to each location
+        for user_detail in self._get_users():
+            username, assigned_location_ids = user_detail
+            if not isinstance(assigned_location_ids, list):
+                assigned_location_ids = [assigned_location_ids]
+            for location_id in assigned_location_ids:
+                if location_id in self._location_details_by_location_id:
+                    self._location_details_by_location_id[location_id]['assigned_users'].append(username)
+
+    def _get_users(self):
+        location_ids = list(self._location_details_by_location_id.keys())
+        # ToDo: confirm that its good to use base_username instead of username from ES, something about
+        #  them not being analysed
+        return UserES().location(location_ids).values_list('base_username', 'assigned_location_ids')
+
+    def _create_workbook(self):
+        wb = Workbook()
+        for location_type, rows in self._create_rows().items():
+            ws = wb.create_sheet(location_type)
+            uniq_headers = self._extract_unique_headers(rows)
+            ws.append(uniq_headers)
+            for row in rows:
+                ws.append([row.get(header) for header in uniq_headers])
+        return wb
+
+    def _create_rows(self):
+        def append_row(username):
```

Comment:
not passing username would mean something like

```
            if location_details['assigned_users']:
                for username in location_details['assigned_users']:
                    append_row()
            else:
                username = ''
                append_row()
```
so this just avoids that `username = ''`

Reply:
What about passing them both as parameters, instead of neither?

Reply:
Ya, that was my question on another place where @esoergel suggested the same, if passing arguments instead of reusing variables from the context of enclosing method which makes me rethink the purpose of having a method defined within a method itself.

Reply:
I admit I don't like nested functions in python, probably because I'm more used to seeing them in javascript, where the scoping rules are slightly different. For this code, my only legitimate complaint is the inconsistency in the two params being different.